### DISCUSSION TITLE
CheckConstraint using postgresql function

### DISCUSSION THREAD

**Discussion by lazydream** [sqlalchemy/sqlalchemy#8568] [Usage Questions]
Hi,

I have a `CheckConstraint` in sqlalchemy that uses PostgreSQL table function created manually beforehand in `alembic` migration, like this:

> class Model(Base):
>         __table_args__ = (
>             CheckConstraint("some_postgresql_function(param_1, param_2)", name="some_constraint"),
>         ) 

The reason to use table function instead of passing the clause in constraint directly is that there is a complex subquery involved as a check so it's better to encapsulate it to a function. 

With that approach I am not able to unit test the business logic related to that ORM model because `Base.metadata.create_all(engine)` is not resolving this check (PostgreSQL function is not created in unit tests since I'm only creating tables here). 

Is there a way or a widely used practice to add such constraint so I can unit test related model without applying `alembic` migrations?

If not, then I'm curious if it is a good practice to apply alembic migrations before unit test?

**Comment by zzzeek** [comment]
if you have that function in your constraint, you'd need to create that as well in the database that you are testing towards. 

for a create_all(), you can put the CREATE FUNCTION code in a DDL event handler, see https://docs.sqlalchemy.org/en/14/core/ddl.html?highlight=ddl#controlling-ddl-sequences .  you'd want to use the before_create event.
